### PR TITLE
fix(price): bound Nostr-sourced price staleness at 1.5x TTL, not 2x

### LINKED ISSUE #860: Nostr price data can be served for ~2x max_price_staleness_seconds due to as_of re-stamp
## Summary

A Nostr-sourced price can be served for roughly twice
`max_price_staleness_seconds` before the staleness check rejects it, because
the same TTL is applied once at ingestion and then the clock is reset.

Two pieces combine:

**1. The provider accepts events up to a full TTL old.**
`src/price/providers/nostr.rs:115` derives its acceptance window from the
shared staleness setting:

```rust
max_age: Duration::from_secs(max_price_staleness_seconds.max(1) as u64),
```

`rank_candidates` (`nostr.rs:193`) then keeps any event whose `created_at` is
within that window:

```rust
.filter(|e| now.as_secs().saturating_sub(e.created_at.as_secs()) <= max_age_secs)
```

**2. The store then re-stamps it as fresh.** `src/price/store.rs:78` writes
`as_of: now` on refresh, discarding the event's own age, and the read path
(`store.rs:102`) measures staleness from that stamp:

```rust
if now.saturating_sub(entry.as_of) <= max_staleness_secs {
```

## Effect

With `max_price_staleness_seconds = 1800`, an event that is already 1799s old
passes the provider's `max_age` gate, gets stamped `as_of = now`, and is then
served for another 1800s before the store considers it stale — roughly 3599s of
total age against a setting whose name implies 1800s.

The setting is doing double duty: it is both "how old an upstream event may be
when we accept it" and "how long we may serve what we accepted". Those are
different questions, and using one number for both compounds them.

Worth noting this is specific to the Nostr provider: the HTTP adapters fetch
current values, so ingestion time and observation time are effectively the same
and `as_of = now` is accurate for them. Nostr is the one source that hands us
data with a meaningful, and potentially old, timestamp of its own.

## Context

This surfaced during review of #841. Closing the provenance item there,
@arkanoider resolved it via the review's alternative rather than full origin
tracking, and noted in writing that a protocol-level origin tag remained a
reasonable follow-up — related to, though not the same as, the re-stamp
behaviour described here. Raising it separately so it is written down rather
than carried in a PR thread.

## Possible direction

The obvious candidate is to carry the event's own `created_at` through to
`as_of` instead of overwriting it with `now`, so staleness is measured from
when the price was observed rather than when we ingested it.

Deliberately not proposing a patch: `as_of` is a shared contract used by every
provider and by the last-known-good preservation logic (`store.rs:64-66`), so
the right shape here is a maintainer call rather than something to decide from
the outside. Happy to implement whichever direction you prefer.

No priority label applied — leaving that to triage.

---

Verified against `upstream/main` @ `7b1118a` on 2026-08-09. Searched open and
closed issues for duplicates — nothing overlapping.

### PR DESCRIPTION
**Pull request by ToRyVand**
Closes #860. Thanks @arkanoider for the ping — took it as requested.

## The bug

`NostrProvider::rank_candidates` (`src/price/providers/nostr.rs`) accepted a trusted-node rate event up to the full `max_price_staleness_seconds` old. `PriceManager::update_all` then stamped every currency in the tick's aggregate with `as_of = now` regardless of source, and the store (`PriceStore::get`) allows serving an entry for another full `max_price_staleness_seconds` before refusing it. Net: a Nostr-sourced price could be served for close to **2x** the configured TTL — against a setting whose name implies one TTL.

## Why not "carry `created_at` through to `as_of`" (the issue's own original suggestion)

That framing was written against an older shape of this code. Today, `Quote` (`src/price/provider.rs`) and `AggregateResult` (`src/price/aggregate.rs`) carry **no timestamp at all** — `aggregate.rs` is explicitly documented as a pure, clockless transform, and `store.update()` takes one `now: i64` for the whole tick's aggregate, not per-source. Threading a real per-source timestamp through `Quote` → `aggregate_tick` → `AggregateResult` → `store` for every provider would be a real architecture change.

It also turns out to be more than the bug needs: Nostr is already restricted to **fallback-only** (`restrict_nostr_to_fallback` in `manager.rs`) — for any currency a non-Nostr provider can cover this tick, Nostr's quote for that currency is dropped before aggregation ever sees it. Nostr's quote only survives to affect `as_of` when it's the **sole** contributor for that currency this tick. Every HTTP-sourced `as_of` is already accurate (ingestion ≈ observation for an HTTP fetch), so the double-staleness bug is real only in the Nostr-exclusive case.

## The fix

Given that narrower blast radius, this closes it entirely inside the Nostr provider's own acceptance gate — no changes to `Quote`, `AggregateResult`, `aggregate_tick`, or `PriceStore`.

New `[price]` setting `nostr_ingestion_budget_pct: f64` (default `0.5`). The Nostr provider's `max_age` becomes `max_price_staleness_seconds × nostr_ingestion_budget_pct` instead of the full TTL — the setting's "how old may an event be when accepted" and "how long may we serve what we accepted" roles get split instead of double-counted. Worst case total age: `0.5×TTL` (ingestion) `+ 1.0×TTL` (store's own unchanged window) `= 1.5×TTL`. Not an exact 1.0x bound (that would need the full timestamp-threading refactor above), but it directly closes the issue's own framing ("served for ~2x") and gives operators a knob to tighten further (e.g. `0.2` → 1.2x) without more code changes. `0.5` as the default balances that against the Nostr provider's purpose as a fallback-of-last-resort — a too-small ingestion window would reject more still-useful, merely-lagging events, worst in exactly the case (Nostr-exclusive currency) where losing the source is worst.

`nostr_anchor_dependent` (fiat-cross-via-Nostr-anchor) currencies are out of scope here — that flag protects a different thing (republication provenance, PR #841). They benefit transitively from a fresher anchor without needing separate handling; noted in the docs rather than silently left unaddressed.

## Changes

- `src/price/config.rs`: `PriceSettings.nostr_ingestion_budget_pct` (default `0.5`), validated to `(0, 1]`.
- `src/price/providers/nostr.rs`: `NostrProvider::new` takes the new param; `max_age` is scaled before the existing zero-floor.
- `src/price/manager.rs`: `build_provider`/`from_settings` thread the new setting through.
- `settings.tpl.toml`, `docs/PRICE_PROVIDERS.md`: documented.

## Test plan

- [x] New regression test proving the fix: an event 1000s old — well inside the *old* unscaled 1800s window — is accepted under the old gate and **rejected** under the new default-scaled 900s gate.
- [x] Budget scaling (`0.25 × 2000 = 500`) and the zero-floor edge case (a budget that would round to 0s) both covered.
- [x] Config validation: budget `<= 0`, `> 1.0` rejected; `1.0` and `0.5` accepted.
- [x] All existing `price::*` tests updated for the new `NostrProvider::new` signature and passing (146 tests in the module).

`cargo build`, `cargo clippy --all-targets -- -D warnings`, `cargo fmt --check`, `cargo test --bin mostrod` (1190 passed), `markdownlint-cli2` on the touched doc — all clean.

## Summary by CodeRabbit

* **New Features**
  * Added a configurable Nostr ingestion freshness budget, defaulting to 50% of the normal price staleness window.
  * Added settings documentation, tuning examples, and guidance on currency-specific behavior.

* **Bug Fixes**
  * Prevented compounded freshness windows by limiting how long Nostr events can be accepted.
  * Added safeguards for very short windows, including a one-second minimum.

* **Validation**
  * Invalid budget values are rejected, while supported boundary values are accepted.

### CODE REVIEW

**Comment by arkanoider** [comment]
Great job @ToRyVand i will be looking into this asap...sorry with lnp2p bot closure work is exploding for a small team like us!

File: src/price/providers/nostr.rs [hunk 1]
```diff
@@ -110,9 +118,13 @@ impl NostrProvider {
...
-            // `PriceSettings::validate` already rejects non-positive values;
-            // clamp here so a zero can never make every event look fresh.
-            max_age: Duration::from_secs(max_price_staleness_seconds.max(1) as u64),
+            // `PriceSettings::validate` already rejects non-positive
+            // staleness/budget values; clamp here so a zero (or a scaled
+            // result that rounds down to zero) can never make every event
+            // look fresh.
+            max_age: Duration::from_secs(
+                ((max_price_staleness_seconds as f64 * nostr_ingestion_budget_pct) as u64).max(1),
+            ),
```
[the remark this answers is not in the document]
Comment (ToRyVand) [reply src/price/providers/nostr.rs:131]: Thanks @arkanoider, no rush — totally understandable with the lnp2p bot closure keeping the team stretched. Just pushed a small follow-up (`d511230`) documenting an edge case coderabbit flagged (the 1-second floor on the ingestion window), so it's ready whenever you get to it.

On a personal note — I'm genuinely motivated to keep contributing to the team wherever it's useful, so count on me for more of this.

**Comment by ToRyVand** [comment]
Thanks @arkanoider, no rush — totally understandable with the lnp2p bot closure keeping the team stretched. Just pushed a small follow-up (`d511230`) documenting an edge case coderabbit flagged (the 1-second floor on the ingestion window), so it's ready whenever you get to it.

**Comment by ToRyVand** [comment]
@arkanoider really sorry to hear about lnp2pBot — that's a hard loss, especially for a project so many people relied on. If Mostro's team capacity is stretched right now, I'm glad to help with whatever's most useful — more reviews, more PRs.

**Review by Catrya — CHANGES_REQUESTED** [review]
Request changes — on the design, not the implementation. The bug is real, the
code is clean, 146 price tests pass, clippy and fmt are clean, and the
`restrict_nostr_to_fallback` blast-radius analysis is correct. My problem is
with the premise and the knob.

**The 2× case isn't the general one.** `rank_candidates` already filters
`!e.is_expired_at(now)` (nostr.rs:194), and this same manager stamps
`expiration = now + min(update_interval_seconds * 2, 3600)` on the events it
publishes (manager.rs:537-538) — 600s at the defaults. So for a trusted node
running this code the binding gate was already 600s, not 1800: real exposure
was 600 + 1800 = 1.33×TTL. The 2× case needs a third-party publisher that
omits the expiration tag (the filter is "if any"). That's a legitimate case to
fix, but neither the PR body nor the new docs mention the expiration filter,
and the regression test uses `signed_event`, which carries no expiration tag —
so it's measuring exactly that third-party case without saying so. This
changes what the right default is.

**The repo already has a canonical answer for "how old may a rate event be",
and it isn't a fraction of the TTL.** It's the expiration this node stamps on
its own events: `2 × update_interval_seconds`. Deriving the ingestion window
from that — or refusing an event with no expiration tag — gives 600s at the
defaults, *tighter* than the 900s here (1.33× vs 1.5×), self-tunes with the
poll cadence, and adds no configuration. A knob whose meaning to the operator
is "pick how far past your own TTL you'd like to go" is a strange thing to
expose permanently.

**And 1.0× is closer than the body suggests.** Threading a timestamp through
`Quote → aggregate_tick → AggregateResult → store` for every provider would
indeed be an architecture change, but both pieces needed for the narrow case
already exist: `AggregateResult.contributors` (aggregate.rs:29) gives
per-currency provenance, and `store.update()` takes an arbitrary `now`, so it
can be called twice:

```rust
let (nostr_only, rest) = aggregates.clone().into_iter()
    .partition(|(_, a)| a.contributors == [ProviderId::Nostr]);
self.store.update(rest, now);
self.store.update(nostr_only, now - nostr_event_age);
```

The only genuinely new plumbing is the provider exposing the age of the event
it used. That closes #860 at an exact 1.0×, also without new config. Worth
ruling out explicitly before a permanent setting goes in.

Minor, if the setting survives:

- _pct names a fraction here (0.5 = 50%) while its immediate neighbour
  outlier_threshold_pct = 5.0 means 5%. Someone following the local
  convention writes 50 and gets a startup error — validation catches it, but
  the name invites the mistake. ..._fraction, or keep _pct and take 50.0.
- 1.0 is accepted and documented as "wasteful but not itself invalid" — it
  is precisely the pre-#860 behaviour. Exclude it, or say that plainly.
- The 1-second floor edge case (a 1s TTL with a 0.001 budget) is spelled out
  in five places for a config nobody will ever write. One line covers it.

**Comment by ToRyVand** [comment]
Thanks @Catrya — you're right, and the problem is upstream of this PR. Closing it.

**The premise was wrong in the issue, not just here.** #860 quotes the age
filter at `nostr.rs:193` and stops one line short of `:194`,
`.filter(|e| !e.is_expired_at(now))`. Corrected on the issue: 1.33x for a
publisher running this code, 2.0x only for a third party that omits the tag.

**Which makes this PR a no-op at the defaults** — 900s ingestion window against
a 600s expiration gate, so the gate never moves. Not worth a permanent operator
setting. Closing rather than rebasing: the commits, docs and test all encode the
2x story, and the real fix touches `manager.rs`/`store.rs`, not this branch.

Two notes, both from the source:

**Your 1.0x sketch is more complete than you argued.** I went looking for the
mixed case that would break `contributors == [ProviderId::Nostr]` and it can't
occur — `restrict_nostr_to_fallback` (`manager.rs:600`) drops Nostr's quote for
any currency another provider covers, so `[Nostr, Yadio]` is unreachable.

The gap is one field over: `nostr_anchor_dependent` (`aggregate.rs:30`, set at
`:211`). An El Toque CUP resolved against Nostr's USD is tagged
`contributors = [ElToque]` — your test at `manager.rs:1294` pins that — so it
lands in `rest` with a fresh `as_of` while embedding a relayed rate.

**`2 x update_interval_seconds` I don't think holds.** That's *our* poll
cadence, not the publisher's: a node publishing every 900s would be permanently
rejected by a consumer polling every 300s. Your other variant — refuse an event
with no expiration tag — does hold, and is the narrow fix for the only case that
was ever 2x.

Minor points all correct (`_pct` naming, `1.0` being the pre-#860 behaviour, the
floor in five places).

On the process note from #842: it applies here too. #860 is also my own issue,
and the PR went up before any of the three maintainers @arkanoider asked had
answered — including his endorsement of the carry-`created_at` approach in the
first reply, which is what you've arrived at independently. That's the fix I'll
build, as a fresh PR.
